File: products-go/internal/api/product_controller.go
package api

import (
	"github.com/gofiber/fiber/v2"
	"github.com/nao99/spring-boot-3-vs-go/products-go/internal/api/model"
	"github.com/nao99/spring-boot-3-vs-go/products-go/internal/domain"
	"strconv"
)

type ProductController struct {
	Service domain.ProductService
}

func (ctrl *ProductController) GetProduct(ctx *fiber.Ctx) error {
	productIdString := ctx.Params("productId")

	productId, err := strconv.ParseUint(productIdString, 10, 32)
	if err != nil {
		return err
	}

	product := ctrl.Service.GetProductById(productId)
	if product == nil {
		return fiber.NewError(fiber.StatusNotFound, "Product not found")
	}

	return ctx.JSON(product)
}

func (ctrl *ProductController) GetProducts(ctx *fiber.Ctx) error {
	products := ctrl.Service.GetProducts()
	return ctx.JSON(products)
}

func (ctrl *ProductController) CreateProduct(ctx *fiber.Ctx) error {
	createDto := &dto.CreateProductDto{}

	err := ctx.BodyParser(&createDto)
	if err != nil {
		return err
	}

	product := ctrl.Service.Create(createDto.Category, createDto.Name, createDto.Description)

	return ctx.JSON(product)
}

File: products-spring-boot/src/main/java/org/ndbs/spring/graalvm/product/api/ProductController.java
package org.ndbs.spring.graalvm.product.api;

import org.ndbs.spring.graalvm.product.api.model.CreateProductDto;
import org.ndbs.spring.graalvm.product.domain.ProductService;
import org.ndbs.spring.graalvm.product.domain.model.Product;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.Set;

/**
 * ProductController class <br>
 *
 * <br>
 *
 * @author Nikolai Osipov <nao99.dev@gmail.com>
 * @since  2022-12-11
 */
@RestController
@RequestMapping("/api/v1/products")
public class ProductController {
    private final ProductService service;

    @Autowired
    public ProductController(ProductService service) {
        this.service = service;
    }

    @GetMapping("/{productId}")
    public Product getProduct(@PathVariable("productId") Long id) {
        try {
            return service.getProductById(id);
        } catch (RuntimeException e) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
    }

    @GetMapping
    public Set<Product> getProducts() {
        return service.getProducts();
    }

    @PostMapping
    public Product createProduct(@RequestBody CreateProductDto createDto) {
        return service.createProduct(createDto);
    }
}
